workflow-ui/core-android/src/main/java/com/squareup/workflow1/ui/container/DialogStack.kt

```diff
@@ -0,0 +1,104 @@
+package com.squareup.workflow1.ui.container
+
+import android.content.Context
+import android.os.Parcel
+import android.os.Parcelable
+import android.os.Parcelable.Creator
+import android.view.View
+import androidx.lifecycle.LifecycleOwner
+import com.squareup.workflow1.ui.ViewEnvironment
+import com.squareup.workflow1.ui.WorkflowUiExperimentalApi
+import com.squareup.workflow1.ui.androidx.WorkflowLifecycleOwner
+import com.squareup.workflow1.ui.container.DialogRunner.KeyAndBundle
+
+/**
+ * Does the work of maintaining a set of [Dialog][android.app.Dialog]s
+ * to reflect lists of [Overlay].
+ */
+@WorkflowUiExperimentalApi
+public class DialogStack(
+  private val context: Context,
+  private val getParentLifecycleOwner: () -> LifecycleOwner?,
+  private val createRectProvider: (Overlay) -> RectProvider
+) {
+  public constructor(
+    view: View,
+    createRectProvider: (Overlay) -> RectProvider = { ViewRectProvider(view) }
+  ) : this(view.context, { WorkflowLifecycleOwner.get(view) }, createRectProvider)
+
+  private var runners: List<DialogRunner<*>> = emptyList()
+
+  public val hasDialogs: Boolean = runners.isNotEmpty()
+
+  public fun update(
+    overlays: List<Overlay>,
+    viewEnvironment: ViewEnvironment,
+    beforeShowing: () -> Unit = {}
+  ) {
+    val overlayViewEnvironments = overlays.mapIndexed { index, _ ->
+      viewEnvironment.withBackStackStateKeyPrefix("[${index + 1}]")
+    }
+
+    val newRunners = mutableListOf<DialogRunner<*>>()
+    for ((i, overlay) in overlays.withIndex()) {
+      val overlayEnvironment = overlayViewEnvironments[i]
+      newRunners += if (i < runners.size && runners[i].canShowRendering(overlay)) {
+        runners[i].also { it.showRendering(overlay, overlayEnvironment) }
+      } else {
+        overlay.toDialogFactory(overlayEnvironment).let { dialogFactory ->
+          DialogRunner(
+            overlay, overlayEnvironment, createRectProvider(overlay), context, dialogFactory
+          ).apply {
+            showRendering(overlay, overlayEnvironment)
+            beforeShowing()
+            show(getParentLifecycleOwner())
+          }
+        }
+      }
+    }
```

Comment:
I wish I could pretend to understand what this class is handling for us but I think i'm lost 😵

Reply:
Rather than explain here I'll add comments, and we can see if they get the job done.

Reply:
@steve-the-edwards PTAL

https://github.com/square/workflow-kotlin/pull/588/files#diff-9c32c5692c4fa791f87816ff6af4ee5e17cbde6276a8932c766891a1349d026fR46